nntrainer/src/databuffer_v2.cpp

```diff
@@ -0,0 +1,325 @@
+        std::lock_guard<std::mutex> buffer_data_avail_lock(buffer_m,
+                                                           std::adopt_lock);
+        std::lock_guard<std::mutex> batched_buffer_data_avail_lock(
+          batched_buffer_m, std::adopt_lock);
+        batched_buffer.pop_front();
+        for (unsigned int idx = 0; idx < batch_size; idx++)
+          buffer.pop_front();
+        avail_buffer_idx -= batch_size;
+
+        if (avail_buffer_idx < 0 || avail_buffer_idx > buffer_len)
+          throw std::logic_error("Buffer overflow error");
+      }
+
+      /** Add new space to the buffer */
+      pushBatchedData();
+      /** Wake up background threads waiting for more space to fill data */
+      buffer_cond_hungry.notify_all();
+      return true;
+    }
+
+    // TODO: optimize to wake up only when full batch is filled by the
+    // background thread
+    /** wait for background thread to fill the data */
+    std::unique_lock<std::mutex> batched_buffer_wait_for_data(batched_buffer_m);
+    buffer_cond_filled.wait(batched_buffer_wait_for_data);
+  }
+
+  throw std::logic_error("Reach unexpected code location");
+}
+
+void DataBuffer_v2::collectData() {
+  bool last_element = false;
+  while (!last_element) {
+    bool buffer_space_available;
+    std::tuple<void **, void **> buffer_elem;
+
+    /**
+     * THREAD_REQUEST_TO_STOP : requested by main thread to stop
+     * THREAD_ERROR : thread is in error state, exit
+     * THREAD_EPOCH_FINISHED : thread finished work, exit
+     */
+    {
+      std::lock_guard<std::mutex> thread_stop(thread_m);
+      if (thread_state != ThreadStates::THREAD_RUNNING) {
+        thread_state = ThreadStates::THREAD_STOPPED;
+        buffer_cond_filled.notify_all();
+        return;
+      }
+    }
+
+    /** check if buffer has space available to load the data */
+    {
+      std::lock_guard<std::mutex> buffer_data_fill_lock(buffer_m);
+      buffer_space_available = buffer_len - avail_buffer_idx - 1 > 0;
+      if (buffer_space_available) {
+        buffer_elem = getNthBufferElement();
+        avail_buffer_idx += 1;
+      }
+    }
+
+    /** if the buffer space is received, then call generator cb and get the data
+     */
+    if (buffer_space_available) {
+      void **input_holder = std::get<0>(buffer_elem);
+      void **label_holder = std::get<1>(buffer_elem);
+
+      int status = generator((float **)input_holder, (float **)label_holder,
+                             &last_element, gen_user_data);
+
+      if (status != ML_ERROR_NONE && !last_element) {
+        /** Update the batched buffer element loaded count */
+        std::lock_guard<std::mutex> batched_buffer_update_lock(
+          batched_buffer_m);
+        auto batched_buffer_elem = getNthBatchedBufferElement();
+        std::get<2>(batched_buffer_elem) += 1;
+        /** notify on buffer produce */
+        buffer_cond_filled.notify_all();
+      } else if (status == ML_ERROR_NONE) {
```

Comment:
if `status != ML_ERROR_NONE` it should go to error handling section not here.

```suggestion
      } else if (status == ML_ERROR_NONE && last_element == true) {
```